shopinvader_sale_coupon/tests/test_cart.py

```diff
@@ -0,0 +1,159 @@
+        # changed in shopinvader's CommonConnectedCartCase.
+        # Start with an empty cart
+        self.cart.order_line.unlink()
+        # Config programs
+        self.code_promotion_program.reward_type = "discount"
+
+    def _generate_coupons(self, program, qty=1):
+        existing_coupons = program.coupon_ids
+        # Create coupons
+        self.env["coupon.generate.wizard"].with_context(active_id=program.id).create(
+            {
+                "generation_type": "nbr_coupon",
+                "nbr_coupons": qty,
+            }
+        ).generate_coupon()
+        # Return only the created coupons
+        return program.coupon_ids - existing_coupons
+
+    def test_immediate_promotion_program(self):
+        # Test case 1 (1 A): Assert that no reward is given,
+        # as the product B is missing
+        self.service.dispatch(
+            "add_item",
+            params={
+                "product_id": self.product_A.id,
+                "item_qty": 1.0,
+            },
+        )
+        self.assertEqual(
+            len(self.cart.order_line),
+            1,
+            "The promo offer shouldn't have been applied as the product B "
+            "isn't in the order",
+        )
+        # Test case 2 (1 A 1 B): Assert that the reward is given,
+        # as the product B is now in the order
+        self.service.dispatch(
+            "add_item",
+            params={
+                "product_id": self.product_B.id,
+                "item_qty": 1.0,
+            },
+        )
+        self.assertEqual(
+            len(self.cart.order_line), 3, "The promo should've been applied"
+        )
+        # Test case 3 (1 B): Assert that the reward is removed when the order
+        # is modified and doesn't match the rules anymore
+        self.service.dispatch(
+            "delete_item",
+            params={
+                "item_id": self.cart.order_line[0].id,
+            },
+        )
+        self.assertEqual(
+            len(self.cart.order_line),
+            1,
+            "The promo reward should have been removed as the rules are not "
+            "matched anymore",
+        )
+        self.assertEqual(self.cart.order_line.product_id.id, self.product_B.id)
+
+    def test_code_promotion_program(self):
+        self.code_promotion_program.promo_code = "promocode"
+        # Buy 1 A + Enter code, 1 A is free
+        self.service.dispatch(
+            "add_item",
+            params={
+                "product_id": self.product_A.id,
+                "item_qty": 1.0,
+            },
+        )
+        self.assertEqual(
+            len(self.cart.order_line),
+            1,
+            "The promo offer shouldn't have been applied as the code hasn't "
+            "been entered yet",
+        )
+        # Enter an invalid code
+        with self.assertRaises(UserError):
+            self.service.dispatch("apply_coupon", params={"code": "fakecode"})
```

Comment:
not mandatory, but next time, when you test an exception use `assertRaisesRegex` to ensure that's the exception you are looking for.  Sometimes it can happen that another exception of the same type is raised which means that something else is wrong.